File: Archiver/app_code/clsCompression.vb
Imports System.IO
'Imports Microsoft.Win32
Imports System.IO.Compression
Public Class clsCompression
    Public Function StrToByteArray(ByVal str As String) As Byte()
        Dim encoding As New System.Text.UTF8Encoding()
        Return encoding.GetBytes(str)
    End Function 'StrToByteArray
    Public Function ByteArrayToStr(ByVal dBytes As Byte()) As String
        Dim str As String
        Dim enc As New System.Text.UTF8Encoding()
        str = enc.GetString(dBytes)
        Return str
    End Function

    Public Function CompressStringToBuffer(ByVal tgtStr As String, ByRef OriginalSize As Double, ByRef CompressedSize As Double) As Byte()

        Dim CompressedDataBuffer() As Byte
        CompressedDataBuffer = StrToByteArray(tgtStr)

        If CompressedDataBuffer.Length = 0 Then
            CompressedDataBuffer = Nothing
            Return CompressedDataBuffer
        End If

        OriginalSize = CompressedDataBuffer.Length

        Dim gzBuffer() As Byte = CompressBuffer(CompressedDataBuffer)

        CompressedDataBuffer = gzBuffer
        CompressedSize = CompressedDataBuffer.Length
        Return CompressedDataBuffer

    End Function

    Public Function DeCompressBufferToString(ByVal CompressedDataBuffer As Byte(), ByRef OriginalSize As Double, ByRef CompressedSize As Double) As String
        CompressedSize = CompressedDataBuffer.Length
        Dim DeCompressedDataBuffer As Byte() = DecompressBuffer(CompressedDataBuffer)
        OriginalSize = DeCompressedDataBuffer.Length
        Dim S As String = ByteArrayToStr(DeCompressedDataBuffer)
        Return S
    End Function

    Public Function CompressBuffer(ByVal BufferToCompress As Byte()) As Byte()
        Dim ms As New MemoryStream()
        Dim zip As New GZipStream(ms, CompressionMode.Compress, True)
        zip.Write(BufferToCompress, 0, BufferToCompress.Length)
        zip.Close()
        ms.Position = 0

        Dim outStream As New MemoryStream()

        Dim compressed As Byte() = New Byte(ms.Length - 1) {}
        ms.Read(compressed, 0, compressed.Length)

        Dim gzBuffer As Byte() = New Byte(compressed.Length + 3) {}
        Buffer.BlockCopy(compressed, 0, gzBuffer, 4, compressed.Length)
        Buffer.BlockCopy(BitConverter.GetBytes(BufferToCompress.Length), 0, gzBuffer, 0, 4)
        Return gzBuffer

    End Function
    Public Function DeCompressBuffer(ByVal BufferToDecompress As Byte()) As Byte()
        Dim ms As New MemoryStream()
        Dim msgLength As Integer = BitConverter.ToInt32(BufferToDecompress, 0)
        ms.Write(BufferToDecompress, 4, BufferToDecompress.Length - 4)

        Dim buffer As Byte() = New Byte(msgLength - 1) {}

        ms.Position = 0
        Dim zip As New GZipStream(ms, CompressionMode.Decompress)
        zip.Read(buffer, 0, buffer.Length)

        Return buffer
    End Function
End Class

File: Archiver/app_code/clsCompression.cs
using System;
using global::System.IO;
// Imports Microsoft.Win32
using global::System.IO.Compression;
using MODI;

namespace EcmArchiver
{
    public class clsCompression
    {
        public byte[] StrToByteArray(string str)
        {
            var encoding = new global::System.Text.UTF8Encoding();
            return encoding.GetBytes(str);
        } // StrToByteArray

        public string ByteArrayToStr(byte[] dBytes)
        {
            string str;
            var enc = new System.Text.UTF8Encoding();
            str = enc.GetString(dBytes);
            return str;
        }

        public byte[] CompressStringToBuffer(string tgtStr, ref double OriginalSize, ref double CompressedSize)
        {
            byte[] CompressedDataBuffer;
            CompressedDataBuffer = StrToByteArray(tgtStr);
            if (CompressedDataBuffer.Length == 0)
            {
                CompressedDataBuffer = null;
                return CompressedDataBuffer;
            }

            OriginalSize = CompressedDataBuffer.Length;
            var gzBuffer = CompressBuffer(CompressedDataBuffer);
            CompressedDataBuffer = gzBuffer;
            CompressedSize = CompressedDataBuffer.Length;
            return CompressedDataBuffer;
        }

        public string DeCompressBufferToString(byte[] CompressedDataBuffer, ref double OriginalSize, ref double CompressedSize)
        {
            CompressedSize = CompressedDataBuffer.Length;
            var DeCompressedDataBuffer = DeCompressBuffer(CompressedDataBuffer);
            OriginalSize = DeCompressedDataBuffer.Length;
            string S = ByteArrayToStr(DeCompressedDataBuffer);
            return S;
        }

        public byte[] CompressBuffer(byte[] BufferToCompress)
        {
            var ms = new MemoryStream();
            var zip = new GZipStream(ms, CompressionMode.Compress, true);
            zip.Write(BufferToCompress, 0, BufferToCompress.Length);
            zip.Close();
            ms.Position = 0L;
            var outStream = new MemoryStream();
            var compressed = new byte[(int)(ms.Length - 1L + 1)];
            ms.Read(compressed, 0, compressed.Length);
            var gzBuffer = new byte[compressed.Length + 3 + 1];
            Buffer.BlockCopy(compressed, 0, gzBuffer, 4, compressed.Length);
            Buffer.BlockCopy(BitConverter.GetBytes(BufferToCompress.Length), 0, gzBuffer, 0, 4);
            return gzBuffer;
        }

        public byte[] DeCompressBuffer(byte[] BufferToDecompress)
        {
            var ms = new MemoryStream();
            int msgLength = BitConverter.ToInt32(BufferToDecompress, 0);
            ms.Write(BufferToDecompress, 4, BufferToDecompress.Length - 4);
            var buffer = new byte[msgLength];
            ms.Position = 0L;
            var zip = new GZipStream(ms, CompressionMode.Decompress);
            zip.Read(buffer, 0, buffer.Length);
            return buffer;
        }
    }
}